server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GeoHashGridValuesSourceBuilder.java

```
@@ -0,0 +1,117 @@
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.geometry.utils.Geohash;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.bucket.geogrid.CellIdSource;
+import org.elasticsearch.search.aggregations.bucket.geogrid.GeoHashGridAggregationBuilder;
+import org.elasticsearch.search.aggregations.support.ValueType;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class GeoHashGridValuesSourceBuilder extends CompositeValuesSourceBuilder<GeoHashGridValuesSourceBuilder> {
+    static final String TYPE = "geohash_grid";
+
+    private static final ObjectParser<GeoHashGridValuesSourceBuilder, Void> PARSER;
+    static {
+        PARSER = new ObjectParser<>(GeoHashGridValuesSourceBuilder.TYPE);
+        PARSER.declareInt(GeoHashGridValuesSourceBuilder::precision, new ParseField("precision"));
+        CompositeValuesSourceParserHelper.declareValuesSourceFields(PARSER, ValueType.NUMERIC);
+    }
+
+    static GeoHashGridValuesSourceBuilder parse(String name, XContentParser parser) throws IOException {
+        return PARSER.parse(parser, new GeoHashGridValuesSourceBuilder(name), null);
+    }
+
+    private int precision = GeoHashGridAggregationBuilder.DEFAULT_PRECISION;
+
+    GeoHashGridValuesSourceBuilder(String name) {
+        super(name);
+    }
+
+    GeoHashGridValuesSourceBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.precision = in.readInt();
+    }
+
```

should we throw an IAE if `format(DocValueFormat)` is used since it is ignored in the build ?

---

Since we don't declare a `format` field in the `PARSER` we throw an unknown field error and the terms and histogram sources don't do this check. 

I can add it if that is the prevailing opinion.

---

That works for rest queries but users of the HLRC can use this builder directly so it would be nice to fail early if they try to set the format ?

---

Ah, I did not know the HLRC had direct access to this. Will definitely update to prevent setting it.